### ISSUE TITLE
MD053 false-positive thrown over Pandoc/GitHub footnotes

### ISSUE THREAD

**Issue by LeCodingWolfie** [DavidAnson/markdownlint#599]
[Rule MD053][MD053] is incorrectly thrown after referencing a link definition over a Pandoc/Github footnote [(see Pandoc's User Guide for syntax)][Pandoc] [(and GitHub's)][GitHub]. 

However, the rule isn't thrown on (indented) lines directly after, i.e. on long footnotes. 
Though, for Pandoc to render a new line on a (long) footnote, a blank line has to be entered within two blocks of text; and creating a new (indented) block that references a link does throw the same rule violation; likely because it's considered by Markdownlint as a [code block][Guide].

<sup>As I'm not gonna bother checking the code:</sup>
From what I can infer, Markdownlint checks if a link is used based on the first character(s) of a line, such as to check if a line represents a code block, and that's why link references aren't checked/found on footnotes.

# Reproduction
```
$ markdownlint -V
0.32.2
```
<sup>[0.32.2 is the latest version for CLI][CLI], [which uses the latest version of the API (i.e. 0.26.2)][API]</sup>
using the following Markdown file:
```markdown
<!-- pandoc.md -->

# Example with Pandoc Footnotes

A sentence with footnotes: [^1] [^2]
A sentence with named footnotes: [^name] [^name2]
A sentence with a link reference: [Pandoc's User Guide][Pandoc]

[^1]: I am a footnote!
[^2]: I reference a [PCWorld][PCW] article!
[^name]: I am a footnote with name!
[^name2]: I am also a named footnote! I also reference the [PCWorld][PCW] article!

[PCW]: https://www.pcworld.com/article/395018/how-much-ram-do-you-need-in-a-laptop.html#4gb-ram
[Pandoc]: https://pandoc.org/MANUAL.html#extension-citations
```
the CLI outputs:
```
$ markdownlint pandoc.md
pandoc.md:12:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "pcw"] [Context: "[PCW]: https://www.pcworld.com..."]
```
as you can see, using a link reference on a sentence doesn't throw the rule.
___
While not using footnotes at all, obviously doesn't throw any error.
<details>
<summary> Just a superfluous example </summary>
<p>

With the file:
```markdown
<!-- no-pandoc.md -->

# Example without Pandoc Footnotes

A sentence that references a [PCWorld][PCW] article.

[PCW]: https://www.pcworld.com/article/395018/how-much-ram-do-you-need-in-a-laptop.html#4gb-ram
```
the CLI doesn't output anything (i.e. any rule violation):
```
$ markdownlint no-pandoc.md 
```

</p>
</details>

On the case of long footnotes (i.e. those that contain multiple lines/blocks of text), attempting to reference a link on a new (indented) line doesn't throw the error. nor on further (indented) lines of text:
```markdown
<!-- long.md -->

# Example with Long Footnotes

A sentence with a long footnotes: [^long] [^longer] [^longest]

[^long]: I am a long footnote!
    I don't do any harm :)
[^longer]: I am a longer footnote. I do reference the [PCWorld][PCW] article.
    I do harm. Though, not here: [Pandoc's User Guide][Pandoc]
[^longest]: I am the longest footnote. I also reference the [PCWorld][PCW] article.

    I am a harmful new block of text: [Rule MD053][Rule]

[PCW]: https://www.pcworld.com/article/395018/how-much-ram-do-you-need-in-a-laptop.html#4gb-ram
[Pandoc]: https://pandoc.org/MANUAL.html#extension-citations
[Rule]: https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md053---link-and-image-reference-definitions-should-be-needed
```
of which CLI outputs:
```
$ markdownlint long.md
long.md:13:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "pcw"] [Context: "[PCW]: https://www.pcworld.com..."]
long.md:15:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "rule"] [Context: "[Rule]: https://github.com/Dav..."]
```
but, as you can see, the new block of text does throw the rule violation, as already explained.
<details>
<summary>Looks like this as a PDF with Pandoc</summary>

![2022-10-05T22:01:46,483395811-05:00](https://user-images.githubusercontent.com/76713363/194205328-ef9c20cf-26af-4fba-955c-df1314b0485a.png)

Like said, the new block text (i.e. blank line) does print a new line on the PDF.

</details>

<sup>Have to say, a lot of reproduction; but, hope it was helpful</sup>

[MD053]: https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md053---link-and-image-reference-definitions-should-be-needed
[Pandoc]: https://pandoc.org/MANUAL.html#extension-citations
[GitHub]: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#footnotes
[CLI]: https://github.com/igorshubovych/markdownlint-cli/releases/tag/v0.32.2
[API]: https://github.com/DavidAnson/markdownlint/releases/tag/v0.26.2
[Guide]: https://www.markdownguide.org/basic-syntax/#code-blocks

**Comment by DavidAnson** [comment]
Thank you for the exceptionally detailed issue! I don't think I realized you could have references within definitions - and I need to understand the rules for footnotes better to see why four spaces doesn't make a code block in your example.

**Comment by LeCodingWolfie** [comment]
> Thank you for the exceptionally detailed issue!

You're welcome, glad my issue does help :smile:

> [...] why four spaces doesn't make a code block in your example.

Basically, if you look at the Pandoc's User Guide on footnotes [(see)][Pandoc], you can see you need to use eight spaces (or two tab spaces, if you will) instead of four to make a code block, similar to how code blocks are created within lists [(see Markdown Guide)][Code]. In fact, in lists, to add a new paragraph to an element, you have to indent the line with four spaces [(see *id.*)][Lists], so a code block has to be indented with four spaces more — similar to how new lines are rendered by Pandoc.

Thus, at the end of the day, it's just about checking if the user is typing within a long footnote, and look for 8 spaces instead of 4 
to create a code block.

By the way, I know how to program on TS/JS, so I could help with that.

[Pandoc]: https://pandoc.org/MANUAL.html#footnotes
[Code]: https://www.markdownguide.org/basic-syntax/#code-blocks-1
[Lists]: https://www.markdownguide.org/basic-syntax/#paragraphs

**Comment by DavidAnson** [comment]
Typically, the rule for an indented code block is any text that is four spaces _more_ than the current alignment. In unordered list, that can be six spaces as shown here: https://spec.commonmark.org/0.30/#example-270

The Pandoc behavior you link to where extra indentation maintains the flow is not something I am familiar with, and I'm not sure there is precedent in the CommonMark spec for that.

**Comment by DavidAnson** [comment]
FYI, extra spaces are not needed in the third link for lists: http://markdown-it.github.io/#md3=%7B%22source%22%3A%22%2A%20This%20is%20the%20first%20list%20item.%5Cn%2A%20Here%27s%20the%20second%20list%20item.%5Cn%5Cn%20%20I%20need%20to%20add%20another%20paragraph%20below%20the%20second%20list%20item.%5Cn%5Cn%2A%20And%20here%27s%20the%20third%20list%20item.%22%2C%22defaults%22%3A%7B%22html%22%3Afalse%2C%22xhtmlOut%22%3Afalse%2C%22breaks%22%3Afalse%2C%22langPrefix%22%3A%22language-%22%2C%22linkify%22%3Atrue%2C%22typographer%22%3Atrue%2C%22_highlight%22%3Atrue%2C%22_strict%22%3Afalse%2C%22_view%22%3A%22html%22%7D%7D

**Comment by DavidAnson** [comment]
This may be a Pandoc vs. CommonMark difference. If so, I use the CommonMark specification in cases where compatibility is not possible. That doesn't seem to be the case here, though.

**Comment by LeCodingWolfie** [comment]
<sup>Besides the fact that Pandoc's is a non-standard flavour of Markdown</sup>

> Typically, the rule for an indented code block is any text that is four spaces more than the current alignment.

<details>
<summary>Ratonale based on the fact that Pandoc creates a new 4-space alignment:</summary>

<sup>To be honest, this is kind of low-resolution thinking, considering the rest of the _picture_</sup>

That's a key statement, from what I can see. In fact, what Pandoc may be doing is that it sets the alignment of a new line with a 4-space indentation, and as you mention, that sets forth that a code block is ruled at four spaces further, totaling eight spaces.

[Taking from the example you shared from CommonMark][CM], the alignment is set forth from the list element definition (i.e.  the `-` character) and the space, which means that, to make a code block, you need to make a 6-space indent; and of which Pandoc follows alike (4-space align. + 4-space indent. for code block)

[Similar to the Markdown-It demo][MarkdownIt], text is written directly after the 2-space alignment; similar to Pandoc, of which text is written after the 4-space indent. I think my point is very obvious

</details>

Thinking it that way, Pandoc doesn't really break the "alignment rule" ; though, syntactically, it may be awkward to start with the an indefinite footnote start [^1], and then set a new 4-space alignment, which may be your point.

> This may be a Pandoc vs. CommonMark difference. If so, I use the CommonMark specification in cases where compatibility is not possible. That doesn't seem to be the case here, though.

To that end of compatibility, I think the linter can check if the "current context" is within a (long) footnote (check where it starts, and where it ends):

<details>
<summary>Here's an example:</summary>

```markdown
I am a normal paragraph.
    > Here's where text starts
[^1]: I'm text on the first line < Here's where it ends
         > Here's where starts
[^long]: I'm other line of text
I am not aligned
neither do I 
```

</details>

Otherwise, I think it can be possible to add an exception or so.

[^1]: i.e. `[^somelongtextthatdoesntreallyendbutdoesfinish]: `

[CM]: https://spec.commonmark.org/0.30/#example-270
[MarkdownIt]: http://markdown-it.github.io/#md3=%7B%22source%22%3A%22%2A%20This%20is%20the%20first%20list%20item.%5Cn%2A%20Here%27s%20the%20second%20list%20item.%5Cn%5Cn%20%20I%20need%20to%20add%20another%20paragraph%20below%20the%20second%20list%20item.%5Cn%5Cn%2A%20And%20here%27s%20the%20third%20list%20item.%22%2C%22defaults%22%3A%7B%22html%22%3Afalse%2C%22xhtmlOut%22%3Afalse%2C%22breaks%22%3Afalse%2C%22langPrefix%22%3A%22language-%22%2C%22linkify%22%3Atrue%2C%22typographer%22%3Atrue%2C%22_highlight%22%3Atrue%2C%22_strict%22%3Afalse%2C%22_view%22%3A%22html%22%7D%7D

**Comment by LeCodingWolfie** [comment]
Sorry for the close-and-open, I accidentally (by pressing the Enter key, instead of Shift + Enter under a code block) sent the last comment and closed it as completed :sweat_smile:

**Comment by LeCodingWolfie** [comment]
By the way, it's good remarking that this is not only behavior from Pandoc, but GitHub does something similar, but not exactly the same, i.e. you can create a new line two ways, either by indenting the new line with two spaces:
```markdown
[^1]: A line
  A new line
```
or just creating a block of text with a 4-space indent:
```markdown
[^note]:
    I am a new block of text
    With a new line as well
```
which I think is weirder, at least compared to Pandoc.

And, using code blocks on footnotes? They don't work on GitHub, but they should theoretically work if GitHub applied the "alignment rule" on footnotes. [^1] [^2]

<details>
<summary>Here's how I tried to indent code blocks on the footnotes, by the way:</summary>

```markdown
[^1]: I am a footnote with 2-space indent rule
  Here a new line
      Am I a code block?

[^2]: 
    I am another footnote, but with 4-space alignment
        Am I also a code block?
```

</details>

[^1]: I am a footnote with 2-space indent rule
  Here a new line
      Am I a code block?

[^2]: 
    I am another footnote, but with 4-space alignment
        Am I also a code block?
